File: kotlin/canton-wallet-sdk/src/main/kotlin/io/github/vsima/canton/wallet/DelegatingSigningDriver.kt
// Copyright (c) 2026 Victor Sima
// SPDX-License-Identifier: Apache-2.0

package io.github.vsima.canton.wallet

import com.daml.ledger.api.v2.CryptoOuterClass

/**
 * The custody hook: a [SigningDriver] whose key operations are delegated to
 * caller-supplied suspend functions. Wrap any external signer — Fireblocks,
 * BitGo, Dfns, an HSM, a remote co-signer — by adapting its SDK calls to the
 * two callbacks; everything above the driver (party onboarding, interactive
 * submission, token-standard writes) works unchanged.
 *
 * The callbacks own the encodings: [publicKeyProvider] must return a
 * Ledger-API-compatible key (see [CryptoOuterClass.SigningPublicKey]) and
 * [signer] a signature whose format/algorithm match that key. For P-256
 * custody backends that emit raw `r‖s` signatures, convert to DER before
 * returning ([CryptoOuterClass.SignatureFormat.SIGNATURE_FORMAT_DER]).
 */
public class DelegatingSigningDriver(
    private val publicKeyProvider: suspend () -> CryptoOuterClass.SigningPublicKey,
    private val signer: suspend (ByteArray) -> CryptoOuterClass.Signature,
) : SigningDriver {

    override suspend fun publicKey(): CryptoOuterClass.SigningPublicKey = publicKeyProvider()

    override suspend fun sign(bytes: ByteArray): CryptoOuterClass.Signature = signer(bytes)
}

File: swift/Sources/CantonWalletKit/DelegatingSigningDriver.swift
// Copyright (c) 2026 Victor Sima
// SPDX-License-Identifier: Apache-2.0

import CantonLedgerAPI
import Foundation

/// The custody hook: a ``SigningDriver`` whose key operations are delegated
/// to caller-supplied async closures. Wrap any external signer — Fireblocks,
/// BitGo, Dfns, an HSM, a remote co-signer — by adapting its SDK calls to
/// the two closures; everything above the driver (party onboarding,
/// interactive submission, token-standard writes) works unchanged.
///
/// The closures own the encodings: `publicKeyProvider` must return a
/// Ledger-API-compatible key and `signer` a signature whose format and
/// algorithm match it. For P-256 custody backends that emit raw `r‖s`
/// signatures, convert to DER before returning.
public struct DelegatingSigningDriver: SigningDriver {
    private let publicKeyProvider: @Sendable () async throws -> Com_Daml_Ledger_Api_V2_SigningPublicKey
    private let signer: @Sendable (Data) async throws -> Com_Daml_Ledger_Api_V2_Signature

    public init(
        publicKeyProvider: @escaping @Sendable () async throws -> Com_Daml_Ledger_Api_V2_SigningPublicKey,
        signer: @escaping @Sendable (Data) async throws -> Com_Daml_Ledger_Api_V2_Signature
    ) {
        self.publicKeyProvider = publicKeyProvider
        self.signer = signer
    }

    public func publicKey() async throws -> Com_Daml_Ledger_Api_V2_SigningPublicKey {
        try await publicKeyProvider()
    }

    public func sign(_ bytes: Data) async throws -> Com_Daml_Ledger_Api_V2_Signature {
        try await signer(bytes)
    }
}
